mlir/lib/Dialect/Rock/Transforms/RockPipeline.cpp

```diff
@@ -0,0 +1,438 @@
+  // will run sequentially
+  //  +t\s+=== 0 ===++=== 1 ===+
+  //  + 0 +== S0  ==++== S3  ==+
+  //  +===+=========++=========+
+  //  + 1 +== S1  ==+
+  //  +===+=========+
+  //  + 2 +== S2  ==+
+  //  +===+=========+
+  // In this case, we reduced the time slots to 3, and we have 2 set of stages
+  // runnning in parallel. Please note that conflicts can only happen between S0
+  // and S3. If we increase II, we get to :
+  //  +t\s+=== 0 ===++=== 1 ===+
+  //  + 0 +== S0  ==++== S2  ==+
+  //  +===+=========++=========+
+  //  + 1 +== S1  ==++== S3  ==+
+  //  +===+=========++=========+
+  // Now we have only two time slots and 2 sets of parallel stages. conflicts
+  // can happen between S0 and S2 and between S1 and S3. The above is capture in
+  // the following algorithm. `t` is the time slot, i.e., the flowing of the
+  // time, and goes from 0 to II. `stageSet` is the set of stages that run in
+  // parallel.
+  for (int t = 0; t < ii; t++) {
+    int iteration = 0;
+
+    // The following stages need to run in parallel
+    SmallVector<rock::StageOp> parallelStages;
+    for (size_t j = t; j < stages.size(); j += ii) {
+      schedule.push_back({stages[j], iteration++});
+      // The only resource that can conflict btween different stages is memory
+      // If there are memory conflicts we can sort them via multibuffers. I.e.,
+      // we can (logically) provide a different buffer for different cycles
+      for (auto otherStage : parallelStages) {
+        auto dependencies = getDependencies(otherStage, stages[j], dag);
+        for (auto [res, type] : dependencies) {
+          multiBuffers[res]++;
+        }
+      }
+      parallelStages.push_back(stages[j]);
+    }
+  }
+}
+
+// Prune a dependency graph taking into account multi-buffers. Since
+// multi-buffers are logically different for each iteration, if the dependency
+// on a multi-buffer spans multiple iteration then it can be pruned
+DagType pruneGraph(DagType dag, DenseMap<rock::StageOp, int> &iterationMap,
+                   DenseMap<rock::GpuAllocOp, int> &factors) {
+  DagType prunedGraph;
+  // Multibuffers have the logical property of being unique for each iteration
+  // of the loop Hence, if we know we are dealing with a multi-buffer and the
+  // dependency concerns two different iteration. In other woeds, if stageA
+  // accesses LDS in iteration i and stageB accesses LDS in iteration j stageA
+  // and stageB have no dependencies as long as i!=j
+  for (auto [sink, edges] : dag) {
+    for (auto [source, deps] : edges) {
+      DenseSet<std::pair<rock::GpuAllocOp, DependencyType>> newDeps;
+      for (auto [alloc, type] : deps) {
+        // Add the dependency only if it's not over a multiBuffer (factor==1) or
+        // if it's over a multibuffer, if sink and source share the same
+        // iteration
+        if (factors[alloc] == 1 ||
+            (factors[alloc] > 1 &&
+             ((iterationMap[sink] % 2) == (iterationMap[source] % 2)))) {
```

Comment:
Yes, sorry, I forgot to update. However, I was planning to replace that condition with a simpler `(iterationMap[sink] == iterationMap[source])`, because in theory if the LDS accesses happen in different logical loops, they should not interfere(right?)